### PR TITLE
Respect accept header on no handler

### LINKED ISSUE #30329: 400 Bad Request always returns plain text
On a cluster with no `x-pack` and version `6.2.4`:

```
POST http://127.0.0.1:9200/_xpack/migration/upgrade/.watches HTTP/1.1
Connection: Keep-Alive
Accept: application/json
Content-Length: 0
Host: 127.0.0.1:9200
```

```
HTTP/1.1 400 Bad Request
content-type: text/plain; charset=UTF-8
content-length: 79

No handler found for uri [/_xpack/migration/upgrade/.watches] and method [POST]
```

Ideally it returns `application/json` as requested by the `Accept` header. This is how we currently return `405` and `406`

```
POST http://127.0.0.1:9200/_blah HTTP/1.1
Connection: Keep-Alive
Accept: application/json
Content-Length: 0
Host: 127.0.0.1:9200
```

```
{"error":"Incorrect HTTP method for uri [/_blah] and method [POST], allowed: [HEAD, DELETE, PUT, GET]","status":405}
```

One could also argue  it should not be a `400` but a `404` if the resource does not exist:

```
GET http://127.0.0.1:9200/_nodes/safdasd/asdasd/sadasd HTTP/1.1
Connection: Keep-Alive
Accept: application/json
Content-Length: 0
Host: 127.0.0.1:9200
```

```
No handler found for uri [/_nodes/safdasd/asdasd/sadasd] and method [GET]
```

### PR DESCRIPTION
**Pull request by jasontedor**
Today when processing a request for a URL path for which we can not find a handler we send back a plain-text response. Yet, we have the accept header in our hand and can respect the accepted media type of the request. This commit addresses this.

Closes #30329

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/rest/RestController.java [hunk 1]
```diff
@@ -401,9 +401,15 @@ private void handleOptionsRequest(RestRequest request, RestChannel channel, Set<
      * Handle a requests with no candidate handlers (return a 400 Bad Request
      * error).
      */
-    private void handleBadRequest(RestRequest request, RestChannel channel) {
-        channel.sendResponse(new BytesRestResponse(BAD_REQUEST,
-            "No handler found for uri [" + request.uri() + "] and method [" + request.method() + "]"));
+    private void handleBadRequest(RestRequest request, RestChannel channel) throws IOException {
+        try (XContentBuilder builder = channel.newBuilder()) {
```
Comment (spinscale) [inline server/src/main/java/org/elasticsearch/rest/RestController.java]: super minor nit, you could use `newErrorBuilder()` to disable filtering

**Review by spinscale — APPROVED** [review]
LGTM, left a nit which can be freely ignored

File: server/src/main/java/org/elasticsearch/rest/RestController.java (hunk 1 above)
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/rest/RestController.java]: Good one.

**Comment by jasontedor** [comment]
Thanks a lot for reviewing @spinscale.

**Pull request merged**